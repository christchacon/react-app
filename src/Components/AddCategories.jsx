import { useState } from "react"

export const AddCategories = ( { onNewCategory } ) => {//setCategories antes
  
    const [ inputCategoryValue, setInputCategoryValue ] = useState( '' );//Para manejar el estado real del componente
    const [ inputCantValue, setInputCantValue ] = useState( '' );//Para manejar el estado real del componente
    

    const inputCategoryChange = (event) => {
        setInputCategoryValue(event.target.value);
    }

    const inputCantChange = (event) => {
        
        // if(/^[0-9]+$/.test(event.target.value)){
        //     setInputCantValue(event.target.value);// /^[0-9]-+$/ /^-?\d+$/cons
        //     console.log(event.target);
        // }else if(event.target.value.length == 0){
        //     setInputCantValue(event.target.value);
        // }
        if(/^[0-9]*$/.test(event.target.value)) setInputCantValue(event.target.value);
    }

    const onSubmit = (event) =>{
        //Enviar datos al component padre
        event.preventDefault();
        // console.log( inputCategoryValue );
        // console.log( inputCantValue );
        //onAddCategory.useState.setCategories( categories.concat( inputValue ) )
        if(inputCategoryValue.trim().length <= 1 || inputCantValue.trim().length <= 0 || isNaN( inputCantValue )) return;
        //setCategories( categories => categories.concat( inputValue ) );
        const criteriaSearch = {
            category: inputCategoryValue.trim(),
            quantity: inputCantValue.trim()
        }
        onNewCategory( criteriaSearch );
        setInputCategoryValue( '' );
        setInputCantValue('');

    }

    const onResetInput = () =>{
        setInputCategoryValue( '' );
        setInputCantValue( '' );
        //console.log( 'OnResetInput' );
    }

    return (
    
    <>
        <form>
        
            <input 
                type="text"
                placeholder="Buscar gifs"
                value={ inputCategoryValue }
                onChange= { (event) => inputCategoryChange( event )} //Se puede usar solo la referencia a la funciÃ³n. es necesario este evento en el input HTML para manejar el evento 
            />
            <input
                type="text"
                placeholder="Cantidad"
                value={ inputCantValue }
                onChange= { (event) => inputCantChange( event )} 
            />
            <button aria-label="btn-submit" onClick={ onSubmit }>Submit</button>
        </form>
        <button aria-label="btn-reset" onClick={ onResetInput }>Reset form</button>
        
    </>
  )
}
