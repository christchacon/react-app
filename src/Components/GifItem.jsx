import Proptypes from "prop-types";

//Functional Component para renderizar el item Gif con sus atributos y estilos propios
export const GifItem = ( { image } ) => {//(props) tambien sirve para recibir el argumento
  return (
    <div id={ image.id } className="card">
      <img src={ image.url } />
      <p>
        { image.title }
      </p>
    </div>
  )
}

GifItem.Proptypes = {
  url: Proptypes.string.isRequired,
  title: Proptypes.string.isRequired
}
