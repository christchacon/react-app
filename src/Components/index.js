//Archivo indice de componentes
export * from "./AddCategories";
export * from "./GifGrid";
export * from "./GifItem";

