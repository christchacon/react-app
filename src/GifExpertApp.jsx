    //Imports
import { useState } from 'react';    
import { GifGrid, AddCategories } from './Components';

    
    //Funtional Component
    export const GifExpertApp = () => {
        //const categoriesArray = ['Dragon Ball'];
        const [categories, setCategories] = useState( [] );
        
        const onAddCategory = ( criteriaCategory ) =>{
            //const newCategory = 'Samuray X';
            //console.log( criteriaCategory );
            if( categories.includes( criteriaCategory )) return;
            //setCategories( categories.concat( newCategory ));
            //console.log( categories );
            setCategories( [ criteriaCategory, ...categories ] );
            //console.log( `Category: ${ categories} `);
        }

        const onResetCategories = () =>{
            setCategories( [] );
        }
    
      //console.log( categories );
        return (
        
        <>{/*Fragmemto*/}
            <h1>GifExpertApp</h1>
            {/*<AddCategories setCategories={ setCategories }/>*/}

            <AddCategories onNewCategory={ ( criteriaCategory ) => onAddCategory( criteriaCategory ) } />{/* props onNewCategory es un funci√≥n que recibe el valor de la categoria */}

            {/*<button aria-label="btn-addNewCategory" onClick={ onAddCategory }>New category</button>*/}
            <div>
                <button aria-label="btn-reset" onClick={ onResetCategories }>Reset categories</button>
                {/* Lista de categories */}
                {
                    categories.map( ( newCategory ) => 
                    //const c = category;   
                        //console.log(`categoria: ${category.category}`);
                        (
                        <GifGrid
                            category={ newCategory }
                            key={ newCategory.category }
                        />
                        ))
                }
            </div>
            
        </>
        
      )
    }
    