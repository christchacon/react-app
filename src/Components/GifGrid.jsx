//Functional component para agrupar los gifs segÃºn categoria recibida.
import { useFetchGifs } from "../hooks/useFetchGifs";
import { GifItem } from "./GifItem";

export const GifGrid = ( {category} ) => {

   
// //    const category = newCategory.map( cat );
// //     console.log(`GifGrid category: ${newCategory.category}`);
// //    console.log(`GifGrid quantity: ${newCategory.quantity}`);
//     console.log(`GifGrid:`);
//     console.log(category);
   const { listImages, isLoading } = useFetchGifs( category );//Hook custom
   console.log( listImages, isLoading );

    return (
        <>
           <h3>{ category.category }</h3>
           {
                isLoading && ( <h2>Loading...</h2> )
           }
           <div className="card-grid">
                {
                    listImages.map( image => (
                        <GifItem 
                            key={ image.id } 
                            image={ image }
                        />
                        //{ ...image } exparse los atributos de la imagen mediante props
                    ))
                }
            </div>
            
            
            {/*<ol>
                { listImages.map( image => (
                    <li key={ image.id }>
                        <div>
                            <h1>{ image.title }</h1>
                            <p>{ image.url }</p>
                            <img src={ image.url } />
                        </div>
                    </li>
                )) }
            </ol>*/}
        </>
    )
}
