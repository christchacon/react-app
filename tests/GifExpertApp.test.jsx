import { fireEvent, render, screen } from "@testing-library/react";
import { GifItem } from "../src/Components";

describe('Text <GifExpertApp />', () => { 
    const image = {
        id: 'id',
        url: 'https://image.com/image',
        title: 'title'
    }
    test('Should Match to Snapchot', () => { 
        const { container } = render( <GifItem image={ image } /> )
        expect(  container ).toMatchSnapshot();
     });

     test('should show image and title', () => { 
        render( <GifItem image={ image } /> )
        screen.debug();
        //expect( screen.getByRole( 'img' ).src ).toBe( image.url );
        const { src } = screen.getByRole( 'img' );
        expect( src ).toBe( image.url );
        expect( screen.getByText( image.title ) ).toBeTruthy();
      })
 })