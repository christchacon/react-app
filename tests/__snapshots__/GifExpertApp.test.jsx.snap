// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text <GifExpertApp /> Should Match to Snapchot 1`] = `
<div>
  <div
    class="card"
    id="id"
  >
    <img
      src="https://image.com/image"
    />
    <p>
      title
    </p>
  </div>
</div>
`;
